File: src/SEBT.Portal.Core/Models/Auth/IdProofingStatus.cs
namespace SEBT.Portal.Core.Models.Auth;

/// <summary>
/// Workflow state of ID proofing for a user.
/// Separate from IAL (Identity Assurance Level) - status tracks process state, IAL tracks assurance achieved.
/// </summary>
public enum IdProofingStatus
{
    /// <summary>
    /// ID proofing has not been initiated.
    /// </summary>
    NotStarted = 0,

    /// <summary>
    /// ID proofing is currently in progress.
    /// </summary>
    InProgress = 1,

    /// <summary>
    /// ID proofing has been completed successfully.
    /// </summary>
    Completed = 2,

    /// <summary>
    /// ID proofing failed or was rejected.
    /// </summary>
    Failed = 3,

    /// <summary>
    /// ID proofing has expired and needs to be renewed.
    /// </summary>
    Expired = 4
}

File: src/SEBT.Portal.Web/src/lib/idProofingStatus.ts
/**
 * Numeric values mirror `SEBT.Portal.Core.Models.Auth.IdProofingStatus` from the API/JWT.
 */
export const IdProofingStatus = {
  NotStarted: 0,
  InProgress: 1,
  Completed: 2,
  Failed: 3,
  Expired: 4
} as const

export type IdProofingStatusNumber = (typeof IdProofingStatus)[keyof typeof IdProofingStatus]

export function isIdProofingStatusCompleted(status: number | null | undefined): boolean {
  return status === IdProofingStatus.Completed
}

/**
 * After OTP, use this when `/auth/status` returned an authenticated session.
 * The user continues ID proofing unless status is {@link IdProofingStatus.Completed}.
 * Any other value (including missing/null claim on that session) routes to proofing.
 * If `login()` returned no session, surface an error instead of calling this.
 */
export function needsIdProofingFlowAfterOtp(status: number | null | undefined): boolean {
  return !isIdProofingStatusCompleted(status)
}

/**
 * Analytics `id_proofed`: true when the workflow completed or a completion timestamp exists
 * (covers legacy/in-flight shapes where status and timestamp may not align).
 */
export function isIdProofedForAnalytics(
  idProofingStatus: number | null | undefined,
  idProofingCompletedAt: number | null | undefined
): boolean {
  return isIdProofingStatusCompleted(idProofingStatus) || Boolean(idProofingCompletedAt)
}
